### ISSUE TITLE
Discuss error handling strategies

### ISSUE THREAD

**Issue by troyraen** [fornax-navo/fornax-demo-notebooks#56]
cc: @bsipocz @jkrick @marziye-jy 

I think it would be useful to have a group discussion about error handling and come up with a coherent strategy for the project. Maybe the next science platform meeting is a good time for it?

We're each handling errors in different ways. IMO, each is totally reasonable within the narrow context that it's being written. But in doing #49, I see that each causes some sort of confusion or extra work when a different person needs to pick it up in a different context. 

Here's my initial thoughts, just to get the conversation going:

To me it makes sense to use different approaches during different stages of development. So I  don't think that everyone needs to follow exactly the same set of rules. But I do think there's probably a small set of simple things we could all do, or allow others to do, that would make things easier for others without overly burdening our own work. My initial list would be:

- If you catch an error, log it or print it. (Seems to me like this alone could ease many of the challenges we currently have.)
- By default, don't catch errors -- just let them raise. In other words don't put in a try-except statement unless/until you actually run into an error that you need to get around.
- Ideally, only catch the specific error(s) that you're having trouble with (e.g., `except KeyError`).
- If it is difficult to figure out which specific error(s) you need to catch, at least use `except Exception` -- don't use a bare `except`.
- Don't put custom errors in the main notebooks. (That's for me 😀... seems like that was confusing and I should at least keep them on the back end, if I use them at all. But I'm interested to hear/discuss, with everything else.)

**Comment by jkrick** [comment]
I literally have no opinion on the best way to handle errors.  I think it is a great idea to have a set of best practices and I will endeavor to follow those.

**Comment by jkrick** [comment]
Maybe once a set of best practices is agreed upon, we can put it on the confluence page.

**Comment by troyraen** [comment]
Hi all. I had been trying to draft up guidelines for error handling based on feedback from a few different people, but I think I jumped the gun by doing it before everyone had a chance to talk things out together. I ended up having separate/parallel conversations with folks and trying to relay info between, but doing a poor job of it. So I’d like to reset the conversation here by clarifying the things that are my own opinions/contributions vs what I’m interpreting from others, and then I’ll step back and let others speak for themselves and discuss (or not, as everyone wants).

These two starter-list items came from me:

> if you catch an error, log it or print it

(that was just an off-the-cuff idea)

> at minimum, use `except Exception` rather than a bare `except`

(I think this is a good rule of thumb since, for example, the former doesn’t interfere with one’s ability to stop the program using _Ctrl-C_.)

@bsipocz expressed to me several times that she’d prefer that we catch specific errors instead of the very broad `Exception`. This was the biggest external (to me) thing that I tried to work into guidelines and relay to others (mostly @jkrick) who had worked on the notebook before me, but I didn’t do a good job of it. So I’ll stop trying to be the translator here and just invite @bsipocz to directly communicate your thoughts about it to others, if you haven’t yet and want to. 
To be clear about _my_ thoughts here, I agree that it’s better to catch a specific error(s) rather than `Exception`, but I don’t have strong feelings about it and I don’t always go to the trouble of doing it. However, this whole conversation is influencing my approach… In particular, I’m trying to do a lot more of this (advice from @bsipocz): “By default, avoid try/except”.  In any case, @bsipocz is a much more experienced developer than I am, and I’m happy to defer and follow any guidelines you’d like to push for this project.

The actual code that I contributed in this area was the two custom errors, and I’d like to clarify why I put them in and also touch base about where they stand now. I put them in because they were the simplest way I could think of to isolate those two try/except blocks and solve (or at least sidestep) some of the concerns that had been raised about them. Wrapping them with custom errors let the main function (and also developers, etc.) catch a specific error in each spot and move on without actually having to care at all about what’s really going on with cutouts or tractor, making development on the rest of the code a little easier. Then separately we could work on narrowing down the set of errors that actually trigger the custom errors without altering/affecting the rest of the code at all. Current status:

- We’ve since worked out the `CutoutError` — it actually wasn’t even getting triggered, so we just got rid of that try/except and custom error completely. The reason I opted for the `CutoutError` in PR #49 instead of just changing that try/except block was that I did a lot of refactoring in that PR and wanted to be able to test my changes (a “refactor” should not change the output, but changing a try/except usually does change the output so I preferred to give it its own PR).
- The `TractorError` is still being triggered by `except Exception`. Personally, I think it’s probably fine for this particular notebook. As I understand it, @jkrick thinks it’s fine scientifically to just skip the object when tractor does fail. I agree with what I think is @bsipocz's view that it would be better to catch something(s) more specific than `Exception`, but I’m not aware of any problems that this is causing, so to me it doesn’t seem worth spending more time on. But again, happy to defer. Also, I’m not overly attached to the custom error and can drop that out if others prefer.

Zooming back out to general error handling strategies… If others still want to have a group discussion and/or develop guidelines then I’m happy to listen and participate. But it doesn’t need to go any further on my account so I’m going to stop driving it forward and let yinz take it from here, or just leave it here, as you like.

Thanks for your patience with me. I’m enjoying working with this group and appreciate your examples of collaboration.
